File: codeforces/857/D.py
import sys
lines = list(map(str.strip, sys.stdin.readlines()))

import math

class Container(object):
	"""Base class of SortedList"""
	def __init__(self, data=None):
		super(Container, self).__init__()
		self.data = data[:] if data else []

	def __repr__(self):
		return '%s(%s)' % (self.__class__.__name__, repr(self.data))
	def __str__(self):
		return str(self.data)
	def __unicode__(self):
		return unicode(self.data)

	def __cast(self, other):
		if isinstance(other, Container): return other.data
		else: return other
	def __lt__(self, other):
		return self.data < self.__cast(other)
	def __le__(self, other):
		return self.data <= self.__cast(other)
	def __eq__(self, other):
		return self.data == self.__cast(other)
	def __ne__(self, other):
		return self.data != self.__cast(other)
	def __gt__(self, other):
		return self.data > self.__cast(other)
	def __ge__(self, other):
		return self.data >= self.__cast(other)
	def __cmp__(self, other):
		return cmp(self.data, self.__cast(other))
	def __hash__(self):
		return hash(self.data)

	def __len__(self):
		return len(self.data)

	def __getitem__(self, key):
		return self.data[key]
	def __setitem__(self, key, value):
		self.data[key] = value
	def __delitem__(self, key):
		del self.data[key]

	def __iter__(self):
		return self.data.__iter__()
	def __reversed__(self):
		return reversed(self.data)

	def __contains__(self, item):
		return item in self.data

from bisect import bisect_left, bisect_right
import bisect

class SortedList(Container):
	def __init__(self, data=None):
		self.data = sorted(data) if data else []

	def __setitem__(self, key, value):
		raise NotImplementedError

	def __contains__(self, value):
		return bisect_right(self.data, value) > 0

	def add(self, value):
		"""Add an item to this list."""
		index = bisect_right(self.data, value)
		self.data.insert(index, value)

	def pop(self, index=None):
		"""Remove and return item at index (default last).
		Raise IndexError if this list is empty."""
		if index is None:
			index = len(self.data) - 1
		return self.data.pop(index)

	def remove(self, value):
		"""Remove first occurrence of value.
		Raise ValueError if not found."""
		index = self.index(value)
		del self[index]

	def clear(self):
		"""Remove all items from this list."""
		self.data = []

	def copy(self):
		"""Return a shallow copy of this list."""
		return self.__class__(self.data[:])

	def count(self, value):
		"""Return number of occurrences of value."""
		i = bisect_left(self.data, value)
		j = bisect_right(self.data, value)
		return j - i

	def index(self, value, start=0, stop=None):
		"""Return first index of value.
		Raise ValueError if not found."""
		if stop is None:
			stop = len(self.data)
		index = bisect_left(self.data, value, start, stop)
		if index != len(self.data) and self.data[index] == value:
			return index
		raise ValueError

from math import log2

def build_sparse_table(arr):
    n = len(arr)
    logn = int(log2(n)) + 3
    st = [[0 for j in range(logn)] for i in range(n)]
    for i in range(n):
        st[i][0] = arr[i]
    for j in range(1, logn):
        for i in range(n - (1 << j) + 1):
            st[i][j] = max(st[i][j - 1], st[i + (1 << (j - 1))][j - 1])
    return st

def query_sparse_table(st, l, r):
    j = int(log2(r - l + 1))
    return max(st[l][j], st[r - (1 << j) + 1][j])

i = 1
while i < len(lines):
    n = int(lines[i])
    a = []
    b = []
    i+=1
    for _ in range(n):
        x, y = map(int, lines[i].split())
        a.append(x)
        b.append(y)
        i+=1
    low = 0
    high = len(a) - 1
    ab = [(x,y) for x, y in zip(a, b)]
    ab.sort()
    a = [x for x, y in ab]
    b = [y for x, y in ab]
    sl = SortedList()
    stmax = build_sparse_table(b)
    # print(b)
    # print(query_sparse_table(stmax, 0, len(b) - 1))
    result = 10**20
    for ii, (x, y) in enumerate(list(zip(a, b))):
        biggest = 1
        idx = bisect.bisect(sl, x)
        candidates = []
        for d in range(-1, 2):
            if idx + d >= 0 and idx + d < len(sl):
                candidates.append(sl[idx + d])
        candidates.sort(key=lambda z: abs(x-z))
        biggestother = candidates[0] if ii == len(a)-1 else query_sparse_table(stmax, ii+1, len(b)-1)
        otherbig = max(biggestother, candidates[0] if candidates else -10**20)
        # print(idx, x, y, biggestother, candidates, otherbig)
        result = min(result, abs(x - otherbig))
        sl.add(y)
    print(result)

File: codeforces/857/D.cpp
// #include  <bits/stdc++.h>
#include <string>
#include <vector>
#include <iostream>
#include <math.h>
#include <algorithm>
#include <cmath>
#include <set>

using namespace std;

#define ll long long

template <class T>
void print_v(vector<T> &v) { cout << "["; for (auto x : v) cout << x << ","; cout << "\b]\n"; }

ll min(ll a,int b) { if (a<b) return a; return b; }
ll min(int a,ll b) { if (a<b) return a; return b; }
ll max(ll a,int b) { if (a>b) return a; return b; }
ll max(int a,ll b) { if (a>b) return a; return b; }
ll gcd(ll a,ll b) { if (b==0) return a; return gcd(b, a%b); }
ll lcm(ll a,ll b) { return a/gcd(a,b)*b; }
string to_upper(string a) { for (int i=0;i<(int)a.size();++i) if (a[i]>='a' && a[i]<='z') a[i]-='a'-'A'; return a; }
string to_lower(string a) { for (int i=0;i<(int)a.size();++i) if (a[i]>='A' && a[i]<='Z') a[i]+='a'-'A'; return a; }
bool prime(ll a) { if (a==1) return 0; for (int i=2;i<=round(sqrt(a));++i) if (a%i==0) return 0; return 1; }


class SparseTable {
public:
    SparseTable(const std::vector<int>& values) : N(values.size()) {
        // Compute the log base 2 of each index
        int logN = static_cast<int>(std::log2(N)) + 1;
        log2_.resize(N+1);
        for (int i = 2; i <= N; i++) {
            log2_[i] = log2_[i/2] + 1;
        }

        // Compute the sparse table
        table_.resize(logN, std::vector<int>(N));
        for (int i = 0; i < N; i++) {
            table_[0][i] = values[i];
        }
        for (int j = 1; j < logN; j++) {
            for (int i = 0; i + (1 << j) <= N; i++) {
                table_[j][i] = std::max(table_[j-1][i], table_[j-1][i+(1<<(j-1))]);
            }
        }
    }

    // Returns the maximum value in the range [l, r]
    int query(int l, int r) const {
        int j = log2_[r-l+1];
        return std::max(table_[j][l], table_[j][r-(1<<j)+1]);
    }

private:
    int N;
    std::vector<int> log2_;
    std::vector<std::vector<int>> table_;
};

int main() {
    int t;
    cin >> t;
    while(t--){
        int n;
        cin >> n;
        vector<pair<int,int>> xs(n);
        for(int i = 0; i < n; i++){
            pair<int,int> p;
            cin >> p.first >> p.second;
            xs[i] = p;
        }
        sort(xs.begin(), xs.end(), [](const auto& lhs, const auto& rhs) {
            if(lhs.first == rhs.first) return lhs.second < rhs.second;
            return lhs.first < rhs.first;
        });
        vector<int> a(n);
        vector<int> b(n);
        for(int i = 0; i < n; i++){
            a[i] = xs[i].first;
            b[i] = xs[i].second;
        }
        SparseTable st(b);
        ll result = 1e18;
        set<int> sl;
        for(int i = 0; i < n; i++){
            ll biggest = 1;
            auto lb = sl.lower_bound(a[i]);
            vector<int> candidates;
            if (lb != sl.end()) {
                candidates.push_back(*lb);
            }
            if(lb != sl.begin()){
                auto prev = lb;
                prev--;
                candidates.push_back(*prev);
            }
            if(lb != sl.end()){
                auto next = lb;
                next++;
                if(next != sl.end()) candidates.push_back(*next);
            }
            // Sort the candidates by the absolute diff between them and a[i]
            sort(candidates.begin(), candidates.end(), [a, i](const auto& lhs, const auto& rhs) {
                return abs(lhs - a[i]) < abs(rhs - a[i]);
            });
            // print_v(candidates);
            ll biggestother = i == a.size() -1 ? candidates[0] : st.query(i+1, n-1);
            ll otherbig = max(candidates.empty() ? -100000 : candidates[0], biggestother);
            result = min(result, abs(a[i] - otherbig));
            // cout << "i: " << i << " result " << result << endl;
            sl.insert(b[i]);
        }
        cout << result << endl;
    }
}
